src/v/storage/mvlog/entry_stream.cc

```diff
@@ -0,0 +1,45 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "storage/mvlog/entry_stream.h"
+
+#include "base/outcome.h"
+#include "base/seastarx.h"
+#include "bytes/iostream.h"
+#include "storage/mvlog/entry.h"
+#include "storage/mvlog/entry_stream_utils.h"
+#include "storage/mvlog/errc.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+
+#include <memory>
+
+namespace storage::experimental::mvlog {
+
+ss::future<result<std::optional<entry>, errc>> entry_stream::next() {
+    auto header_buf = co_await read_iobuf_exactly(
+      stream_, packed_entry_header_size);
+    if (header_buf.size_bytes() != packed_entry_header_size) {
+        co_return std::nullopt;
+    }
+
+    auto entry_hdr = entry_header_from_iobuf(std::move(header_buf));
+    auto computed_crc = entry_header_crc(entry_hdr.body_size, entry_hdr.type);
```

Comment:
Perhaps I missed the high level details for this in the RFC, but is there always going to be CRC validation when `next()` is called? Or is there any good reason we should allow the caller to opt out of this step?

Reply:
>is there always going to be CRC validation when next() is called

I don't see this not being the case without good reason. The CRC is helpful in ensuring that at least the size in the header is correct (which then helps us skip to the next entry if desired).